Channel upgradability requirements (#3240)
cosmos/ibc-go: docs/requirements/channel-upgradability-requirements.md

# Business requirements

Rather than create a new channel to expand upon the capabilities of an existing channel, channel upgradability enables new features and capabilities to be added to existing channels.

## Problem

> Currently, once a channel is opened and the channel handshake is complete, you cannot change or renegociate the semantics of that channel. This means that if you wanted to make a change to a channel affecting the semantics on both channel ends, you would need to open a new channel, meaning all previous state in the prior channel would be lost. This is particularly important for channels using the ICS 20 (fungible token transfer) application module because tokens are not fungible between channels.

Reviewer: ```suggestion
Currently, once a channel is opened and the channel handshake is complete, you cannot change or renegotiate the semantics of that channel. This means that if you wanted to make a change to a channel affecting the semantics on both channel ends, you would need to open a new channel, meaning all previous state in the prior channel would be lost. This is particularly important for channels using the ICS 20 (fungible token transfer) application module because tokens are not fungible between channels.
```

Upgrading a channel enables upgrading the application module claiming the channel, where the upgrade requires a new packet data structure or adding a middleware at both channel ends. Currently it is possible to make changes to one end of a channel that does not require the counterparty to make changes, for example adding rate limiting middleware. Channel upgradability is solving the problem when changes need to be agreed upon on both sides of the channel.

## Objectives

To enable existing channels to upgrade the application module claiming the channel or add middleware to both ends of an existing channel whilst retaining the state of the channel.

## Scope

A new `ChannelEnd` interface is defined after a channel upgrade, the scope of these upgrades is detailed in the table below.

| Features  | Release |
| --------- | ------- |

> | Performing a channel upgrade results in an application module changing from v1 to v2, claiming the same `channelID` and `portID` | v8.1 |

Reviewer: app module version going from v1 to v2 is not always true, slightly unclear whether its meant that this happens every time or is just a possibility that may happen on a particular upgrade

Reviewer 2: It's an example for an upgrade that may happen.

Reviewer: ah, fair enough! thanks

| Performing a channel upgrade results in a channel with the same `channelID` and `portID` modifying the `connectionHops` | v1 |

# User requirements

## Use cases

Sample use cases:

- Upgrading an existing application module from v1 to v2, e.g. new features could be added to the existing ICS 20 application module which would result in a new version of the module.
- Adding middleware on both sides of an existing channel, e.g. relayer incentivisation middleware, ICS 29, requires middleware to be added to both ends of a channel to incentivise the `recvPacket`, `acknowledgePacket` and `timeoutPacket`.

Upgrades can be initiated in two possible ways:

- __Permissionless__: upgrade initiation is approved by governance or group proposal and a relayer can submit `MsgChannelUpgradeInit`.
- __Permissioned__: upgrade initiation is approved by governance or group proposal and `MsgChannelUpgradeInit` is automatically submitted on proposal passing.

The upgrade approval process can be represented in terms of the following parameters:

> - The set of port ID, channel ID pairs that specify the channels that can be upgraded. It can be possible to specifit a single channel or a set of channels satisfiying certain port ID, channel ID conditions. Examples: upgrading can be approved for __a)__ channels with port ID, channel ID combinations (`transfer`, `channel-0`) & (`transfer`, `channel-1`), __b)__ all channels for port ID `transfer`, __c)__ all channels with port ID starting with `icacontroller` (so that all ICA channels on a controller chain, regardless of the owner address appended to the port ID, can be upgraded).

Reviewer 3: ```suggestion
- The set of port ID, channel ID pairs that specify the channels that can be upgraded. It can be possible to specify a single channel or a set of channels satisfying certain port ID, channel ID conditions. Examples: upgrading can be approved for __a)__ channels with port ID, channel ID combinations (`transfer`, `channel-0`) & (`transfer`, `channel-1`), __b)__ all channels for port ID `transfer`, __c)__ all channels with port ID starting with `icacontroller` (so that all ICA channels on a controller chain, regardless of the owner address appended to the port ID, can be upgraded).
```

> - The parameters that are accepted for the upgrade. Currently they are channel version, channel ordering and connection hops. These parameters should be specified per specific pairs of port ID, channel ID (e.g. for channel (`transfer`, `channel-0`)) or for all channels using a certain port ID (e.g. for all channels on port ID `transfer`). The reason is that it could lead to errors specifying the upgrade parameters for channels on two different ports (e.g. for channels on port ID `transfer` and channels with port ID starting with `icacontroller` it would not make sense to upgrade to verions `ics20-2`, since this version only make sense for transfer channels, but not for interchain accounts channels). If an upgradable parameter is omitted, then we shall assume that the parameter must stay the same after the upgrade.

Reviewer 3: ```suggestion
- The parameters that are accepted for the upgrade. Currently they are channel version, channel ordering and connection hops. These parameters should be specified per specific pairs of port ID, channel ID (e.g. for channel (`transfer`, `channel-0`)) or for all channels using a certain port ID (e.g. for all channels on port ID `transfer`). The reason is that it could lead to errors specifying the upgrade parameters for channels on two different ports (e.g. for channels on port ID `transfer` and channels with port ID starting with `icacontroller` it would not make sense to upgrade to versions `ics20-2`, since this version only make sense for transfer channels, but not for interchain accounts channels). If an upgradable parameter is omitted, then we shall assume that the parameter must stay the same after the upgrade.
```

Some concrete examples to illustrate the above:

- Approve upgrading channel (`transfer`, `channel-0`) to version `ics20-2` until block height 1000.
- Approve upgrading channels (`transfer`, `channel-0`) & (`transfer`, `channel-1`) to version `ics20-2` until block height 1000 AND channel (`icacontroller-cosmos13fdx64cc9afrsnjdytk6drnj3mzdku5qsemlz5`, `channel-2`) to ordering `UNORDERED` until time 1682288559.
- Approve upgrading all channels on port ID `transfer` to version `{"fee_version":"ics29-1","app_version":"ics20-1"}` until block height 1050.
- Approve upgrading all channels with port ID starting with `icacontroller` to version `{"fee_version":"ics29-1","app_version":"ics27-1"}` until block height 1080.

Once an approved upgrade succeeds, it is required that a new governance or group proposal passes to approve a new upgrade to the channel.

Assuming that we have two chains, and that both chains approve a compatible upgrade (i.e. the proposed upgrade parameters are identical on both chains and a channel between them is in the upgrade scope), then the following upgrade initiation combinations are possible:

- Permissionless upgrade initiation on both chain A and B.

>      In this scenario both chains approve a channel upgrade and a relayer can submit `MsgChannelUpgradeInit` on one side to start the handshake witin the timeout period specified in the upgrade approval proposal. Upon successful execution of `MsgChannelUPgradeInit` a relayer can submit `MsgChannelUpgradeTry` on the counterparty.

Reviewer 3: ```suggestion
     In this scenario both chains approve a channel upgrade and a relayer can submit `MsgChannelUpgradeInit` on one side to start the handshake within the timeout period specified in the upgrade approval proposal. Upon successful execution of `MsgChannelUPgradeInit` a relayer can submit `MsgChannelUpgradeTry` on the counterparty.
```

- Permissionless upgrade on chain A and permissioned upgrade on chain B.
- Permissioned upgrade on chain A and permissionless upgrade on B.

>     In these 2 scenarios, both side approve a channel upgrade, but one chain automatically initiates the upgrade process by executing `MsgChannelUpgradeInit` and the other side needs a relayer to submit either `MsgChannelUpgradeInit` or `MsgChannelUpgradeTry`. If `MsgChannelUpgradeInit` is executed on both chains, then we would have a crossing hello uld be that a crossing hello situation occurs if relayer(s) submit `MsgChannelUpgradeInit` on both chains.

Reviewer 3: ```suggestion
    In these 2 scenarios, both sides approve a channel upgrade, but one chain automatically initiates the upgrade process by executing `MsgChannelUpgradeInit` and the other side needs a relayer to submit either `MsgChannelUpgradeInit` or `MsgChannelUpgradeTry`. If `MsgChannelUpgradeInit` is executed on both chains, then we would have a crossing hello uld be that a crossing hello situation occurs if relayer(s) submit `MsgChannelUpgradeInit` on both chains.
```

in addition, I think this sentence needs some clarity?

> then we would have a crossing hello uld be that a crossing hello situation occurs if relayer(s) submit `MsgChannelUpgradeInit` on both chains.

Reviewer 2: Yes :) That sentence was a complete mess :D
Thanks for catching all the grammar mistakes!

- Proposal on chain A passes after the counterparty upgrade timeout specified on chain B, then upgrade will not go through and state on chain B will be rolled back to `OPEN`. If the upgrade initiation timeout has not passed yet on chain B, then a relayer can submit a new `MsgChannelUpgradeInit` on chain A with a new counterparty upgrade timeout.

### Permissioned upgrade on chain A and B

Normal flow:

1. Proposal is submitted on chain A to allow upgrade and initiate upgrade for port ID `transfer`, channel ID `channel-0`.
2. Proposal is submitted on chain B to allow upgrade and initiate upgrade for port ID `transfer`, channel ID `channel-0`.
3. Proposal passes on chain A and:
    a. it allows port ID `transfer`, channel ID `channel-0` to be upgraded;
    b. it executes `MsgChannelUpgradeInit` to upgrade channel version from `ics20-1` to `{"fee_version":"ics29-1","app_version":"ics20-1"}`.
4. Proposal passes on chain B and:
    a. it allows port ID `transfer`, channel ID `channel-0` to be upgraded;
    b. it executes `MsgChannelUpgradeInit` to upgrade channel version from `ics20-1` to `{"fee_version":"ics29-1","app_version":"ics20-1"}`.

> 5. Relayer detects execution of `MsgChannelUpgradeInit` on chain A and submits `MsgChannelUpgradeTry` on chain B proposing to upgrade channel version to `{"fee_version":"ics29-1","app_version":"ics20-1"}`. Since chain B has already executed `MsgChannelUpgradeInit`, the we are in a crossing hello scenario.

Reviewer 3: ```suggestion
5. Relayer detects execution of `MsgChannelUpgradeInit` on chain A and submits `MsgChannelUpgradeTry` on chain B proposing to upgrade channel version to `{"fee_version":"ics29-1","app_version":"ics20-1"}`. Since chain B has already executed `MsgChannelUpgradeInit`, we are in a crossing hello scenario.
```

# Business requirements

Rather than create a new channel to expand upon the capabilities of an existing channel, channel upgradability enables new features and capabilities to be added to existing channels. 

## Problem

> IBC is designed so that a specific application module will claim the capability of a channel. Currently, once a channel is opened and the channel handshake is complete, you cannot change the application module claiming that channel. This means that if you wanted to upgrade an existing application module or add middleware to both ends of a channel, you would need to open a new channel with these new modules meaning all previous state in the prior channel would be lost. This is particularly important for channels using the ICS 20 (fungible token transfer) application module because tokens are not fungible between channels.

Author: I updated the problem description with your feedback, apologies for the delay

Adding middleware on both sides of an existing channel, e.g. relayer incentivisation middleware, ICS 29, requires middleware to be added to both ends of a channel to incentivise the `recvPacket`, `acknowledgePacket` and `timeoutPacket`.

# Functional requirements

## Assumptions and dependencies

- Functional relayer infrastructure is required to perform a channel upgrade.
- Chains wishing to successfully upgrade a channel must be using a minimum ibc-go version in the v8 line.
- Chains proposing an upgrade must have the middleware or application module intended to be used in the channel upgrade configured. 

## Features

### 1 - Configuration

| ID | Description | Verification | Status | 
| -- | ----------- | ------------ | ------ | 
| 1.01 | The parameters for the permitted type of upgrade can be configured on completion of a successful governance proposal or using the `x/group` module  | ------------ | Drafted | 
| 1.02 | A type of upgrade can be permitted for all channels with a specific `portID` or for a subset of channels using this `portID` | ------------ | Drafted |

> | 1.03 | A chain may choose to permit all channel upgrades from counterparties by default | ------------ | Drafted |  

Reviewer 3: should we open an issue to keep track of these potential e2e tests?

> | 1.03 | A chain may choose to permit all channel upgrades from counterparties by default | ------------ | Drafted |  

Reviewer 2: Yes, we'll definitely open one when we start getting closer to the moment we can start writing the e2e tests.
